### PR TITLE
Attribute completion on simple expressions

### PR DESCRIPTION
**Pull request by thomasballinger**
This implements the feature initially described in issue #37, autocompletion of literals, and includes autocompletion of literals as discussed later in the thread.

This implementation is relatively conservative: it will only evaluate lists, tuples, strs, unicodes, dicts. It evaluates no methods called with method call syntax, only `__getitem__` invoked by `[]` on these types.

To find an expression to evaluate before current attribute being completed, ast.parse is called on strings of increasing size - this could be a performance issue.

### CODE REVIEW

File: bpython/simpleeval.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# encoding: utf-8
+"""simple evaluation of side-effect free code
+
+In order to provide fancy completion, some code can be executed safely.
+
+"""
+
+from ast import *
```
Comment (sebastinas) [inline bpython/simpleeval.py]: Please don't. See PEP-8 about wildcard imports.

File: bpython/simpleeval.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    """
+    Safely evaluate an expression node or a string containing a Python
+    expression.  The string or node provided may only consist of:
+    * the following Python literal structures: strings, numbers, tuples,
+        lists, dicts, booleans, and None.
+    * variable names causing lookups in the passed in namespace or builtins
+    * getitem calls using the [] syntax on objects of the types above
+    * getitem calls on subclasses of the above types if they do not override
+        the __getitem__ method and do not override __getattr__ or __getattribute__
+        (or maybe we'll try to clean those up?)
```
Comment (thomasballinger) [inline bpython/simpleeval.py]: This last bullet point is wrong, this isn't implemented

Comment (thomasballinger) [inline bpython/simpleeval.py]: also should mention this is based on ast.literal_eval

File: bpython/test/test_line_properties.py [hunk 3]
```diff
@@ -7,7 +7,8 @@
     current_method_definition_name, current_single_word, \
     current_string_literal_attr, current_indexed_member_access_identifier, \
     current_indexed_member_access_identifier_with_index, \
```
Comment (thomasballinger) [inline bpython/test/test_line_properties.py]: current_indexed_member_access_identifier, current_indexed_member_access_identifier_with_index, and current_indexed_member_access_member are all no longer used and should be deleted.

File: bpython/test/test_autocomplete.py [hunk 4]
```diff
@@ -268,20 +268,20 @@ def __getattr__(self, attr):
...
+class TestExpressionAttributeCompletion(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.com = autocomplete.ArrayItemMembersCompletion()
+        cls.com = autocomplete.ExpressionAttributeCompletion()
 
     def test_att_matches_found_on_instance(self):
         self.assertSetEqual(self.com.matches(5, 'a[0].', locals_={'a': [Foo()]}),
-                            set(['a[0].method', 'a[0].a', 'a[0].b']))
+                            set(['method', 'a', 'b']))
```
Comment (sebastinas) [inline bpython/test/test_autocomplete.py]: Why is this changed?

Comment (thomasballinger) [inline bpython/test/test_autocomplete.py]: It's changed because this implementation (unlike TestArrayItemCompletion) never calculates what portion of the line the expression corresponds to. This should be alright because `Completer.matches()` is supposed to replace the portion of the line that its completer's `locate()` returns. ExpressionAttributeCompletion's `locate()` method returns only the attribute, like DictKeyCompletion.

File: bpython/test/test_line_properties.py [hunk 5]
```diff
@@ -224,6 +225,7 @@ def test_simple(self):
         self.assertAccess('stuff[asdf[asd|fg]')
         self.assertAccess('Object.attr1.<|attr2>')
         self.assertAccess('Object.<attr1|>.attr2')
+        self.assertAccess('Object.<attr1|>.attr2')
```
Comment (thomasballinger) [inline bpython/test/test_line_properties.py]: duplocated line

File: bpython/simpleeval.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        elif isinstance(node, List):
+            return list(map(_convert, node.elts))
+        elif isinstance(node, Dict):
+            return dict((_convert(k), _convert(v)) for k, v
+                        in zip(node.keys, node.values))
+        elif isinstance(node, Name):
+            try:
+                return namespace[node.id]
+            except KeyError:
+                return __builtins__[node.id]
```
Comment (thomasballinger) [inline bpython/simpleeval.py]: This is an implementation detail that breaks in pypy, import builtins instead

**Comment by thomasballinger** [comment]
`class StringLiteralAttrCompletion(BaseCompletionType)` should also be removed and its test cases converted as its just a special case of this completion.

File: bpython/simpleeval.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        elif isinstance(node, BinOp) and \
+             isinstance(node.op, (Add, Sub)) and \
+             isinstance(node.right, Num) and \
+             isinstance(node.right.n, complex) and \
+             isinstance(node.left, Num) and \
+             isinstance(node.left.n, (int, long, float)):
```
Comment (sebastinas) [inline bpython/simpleeval.py]: What the purpose of this block?

Comment (thomasballinger) [inline bpython/simpleeval.py]: Complex numbers literals. I should have noted that this is mostly the same as the stdlib ast.literal_eval

Comment (sebastinas) [inline bpython/simpleeval.py]: Then please add a comment. The current combination of types here is confusing. Also, what's the expected result for `d[1+1j].` and `d[1j].` for `d = {1+1j: "string", 1j: "string"}`? For `d[1j].` I get member completion, while I don't get it for `d[1+1j]`.

Note that according to 2.4.3 of the Python 3 documentation there are no complex number literals (only integer, float and imaginary literals).

Comment (thomasballinger) [inline bpython/simpleeval.py]: Hm, working in Python 2 but not 3, so looks like you've hit exactly on the problem. I'll look at how much code can be shared between the two, they might have to be separate implementations.

File: bpython/simpleeval.py (hunk 2 above)
Comment (thomasballinger) [inline bpython/simpleeval.py]: Add tests for exactly these capabilities - I didn't realize complex numbers broke in Python 3 because there weren't tests for it.

**Comment by sebastinas** [comment]
Since `bpython/simpleeval.py` contains plenty of code from Python's `ast` module, the file needs proper attribution and a copyright header.

**Comment by thomasballinger** [comment]
Can I use the license header from bpython/autocomplete for this? Should it instead reference to the Python License?

**Comment by thomasballinger** [comment]
Thoughts on whether it's preferable to keep the simple_eval function single-source and drop the complex number literals, or make two versions that more closely mirror the corresponding ast.literal_eval implementations? This function might get more complicated in the future if the evaluation becomes less conservative (e.g. allow some attribute lookup on namedtuples).

**Comment by thomasballinger** [comment]
Todo: when methods are completed they don't have function docstrings and parameters, probably because current function code can't get the function object.

**Comment by thomasballinger** [comment]
@sebastinas I've added the Python license. I'd be happy to throw this code out and start again if we don't want this code in here.

Otherwise, this still needs more tests.

**Comment by thomasballinger** [comment]
Also, should I add (c) bpython authors and then make an exception for those two functions?

**Comment by thomasballinger** [comment]
Todo
* test that methods completed from expressions show docstrings, with the cursor at various places in the 
* factor out word.startswith('__') behavior
* test that custom number types won't have method run in simple_eval
* test that unary and binary operations work on numbers in simple_eval
* improve evaluate_current_expression docstring

File: bpython/simpleeval.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
```
Comment (sebastinas) [inline bpython/simpleeval.py]: Thanks.

File: bpython/test/test_autocomplete.py (hunk 4 above)
Comment (sebastinas) [inline bpython/test/test_autocomplete.py]: Okay.

File: bpython/simpleeval.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        node_or_string = node_or_string.body
+
+    string_type_nodes = (ast.Str, ast.Bytes) if py3 else (ast.Str,)
+    numeric_types = (int, float, complex) + (() if py3 else (long,))
+
+    # added in Python 3.4
+    if hasattr(ast, 'NameConstant'):
+        name_type_nodes = (ast.Name, ast.NameConstant)
+    else:
+        name_type_nodes = (ast.Name,)
```
Comment (sebastinas) [inline bpython/simpleeval.py]: `string_type_nodes`, `numeric_types`, `name_type_nodes` all look like the could be global.

File: bpython/simpleeval.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        if isinstance(value, list):
+            for item in value:
+               if isinstance(item, ast.AST):
+                    r = find_attribute_with_name(item, name)
+                    if r:
+                        return r
+        elif isinstance(value, ast.AST):
+            r = find_attribute_with_name(value, name)
+            if r:
+                return r
```
Comment (sebastinas) [inline bpython/simpleeval.py]:
```
def find_attribute_with_name(node, name):
    if isinstance(node, ast.Attribute) and note.attr == nane:
        return node
    for item in ast.iter_child_nodes(node):
        r = find_attribute_with_name(item)
        if r:
            return r
```
should do the same and is not a code copy.

**Comment by thomasballinger** [comment]
Fixed, I'm tempted to do the same thing with the other function. I've learned my lesson, I won't copy code unnecessarily in the future.

**Comment by sebastinas** [comment]
There seem to be conflicts. Could you rebase the branch on top of the current master?
